include/oneapi/dpl/pstl/utils.h

```diff
@@ -567,6 +576,62 @@ struct __lifetime_keeper : public __lifetime_keeper_base
     __lifetime_keeper(Ts... __t) : __my_tmps(::std::make_tuple(__t...)) {}
 };
 
+//-----------------------------------------------------------------------
+// Generic bit- and number-manipulation routines
+//-----------------------------------------------------------------------
+
+// Bitwise type casting, same as C++20 std::bit_cast
+template <typename _Dst, typename _Src>
+::std::enable_if_t<
+    sizeof(_Dst) == sizeof(_Src) && ::std::is_trivially_copyable_v<_Dst> && ::std::is_trivially_copyable_v<_Src>, _Dst>
+__dpl_bit_cast(const _Src& __src) noexcept
+{
+#if __cpp_lib_bit_cast >= 201806L || _ONEDPL___cplusplus >= 202002L && __has_include(<bit>)
+    return ::std::bit_cast<_Dst>(__src);
+#elif _ONEDPL_BACKEND_SYCL && _ONEDPL_LIBSYCL_VERSION >= 50300
+    return sycl::bit_cast<_Dst>(__src);
+#elif __has_builtin(__builtin_bit_cast)
+    return __builtin_bit_cast(_Dst, __src);
+#else
+    _Dst __result;
+    ::std::memcpy(&__result, &__src, sizeof(_Dst));
+    return __result;
+#endif
+}
+
+// The max power of 2 not exceeding the given value, same as C++20 std::bit_floor
+template <typename _T>
+::std::enable_if_t<::std::is_integral<_T>::value && ::std::is_unsigned<_T>::value, _T>
+__dpl_bit_floor(_T __x) noexcept
+{
+    if (__x == 0)
+        return 0;
+#if __cpp_lib_int_pow2 >= 202002L || _ONEDPL___cplusplus >= 202002L && __has_include(<bit>)
+    return ::std::bit_floor(__x);
+#elif _ONEDPL_BACKEND_SYCL
+    // Use the count-leading-zeros function
+    return 1 << (sycl::clz(_T{0}) - sycl::clz(__x) - 1);
```

Comment:
is not is better to use:

```suggestion
    return 1 << (sizeof(_T)) - sycl::clz(__x) - 1);
```

Reply:
`sizeof` returns the number of bytes, while we need the number of bits. I'd need to include `<numeric_limits>` etc. to make it right. Is it worth it? I guess `clz` maps efficiently to the hardware.

Reply:
> I guess clz(0) can be optimized by compilers and/or maps efficiently to the hardware.

Probably but it's still runtime. Why should we rely on the optimization and "constructing" initialized object when all we need is know at compile-time.

Why `<limits>`? Probably just `sizeof(T) * 8`?